### PR TITLE
[READY] Fishing and aquarium expansion.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
So, fishing is a feature I enjoyed the few times I played it, though it has its shortcomings and issues, and a few peeves and ideas I've had with it. There's a lot that I've done in this PR, so much that, if I knew earlier it would have been so many lines of code, I'd have split this in several other PRs. It's been a hot while since I've got this carried away. Sorry! I might try to atomize a few things later to make it easier to review.

Listing the changes, off the top of my head:
- Resprited fishing rods, hooks, and the worm bait!
- Added a new, telescopic fishing rod, that can be bought as a goodie. The master rod is also telescopic now.
- Added a couple hooks. One that lets you move the bait up and down, otherwise keeping it in place, and another that stops the fish from escaping, but slowly kills it. The former from the bepis fishing tech node, the latter frm the black market.
- Added a fishing skill and relative legendary reward: A fishing hat, like the one that recites "women fear me, fish fear me"
- You can now stop fishing by activating the fishing rod in your hand, and stops it from stealing all clicks on other things if it isn't in your active hand.
- Reworked fishing traits into fish traits, which can apply to fish after it has been caught. 
- Expanded the fish breeding system. Traits may be passed down to offsprings, and offsprings may evolve (mutate?) into different kind of fishes if conditions when conditions are met.
- Added half a dozen new fishes, each with its own traits: lubefish, sludgefish (and its purple variant), slimefish, unmarine bonemass and unmarine mastodon. Also, holodeck fish, as a joke.
- New traits: lubed skin, parthenogenesis, toxic (new reagent), toxin immunity, predator, necrophage, no mating, crossbreeder, aggressive and revival. Converted Emulsijack's ability and Donkfish's yuckiness into traits as well.
- Added a fish analyzer that you can scan aquariums and fishes with.
- Fish can now be blended if you really want to. The number of reagents from blending, w_class, and the number of fillets you get from cutting fish now scale with size and weight.
- fish feed is no longer infinite (but it should still be plenty).
- Implemented temperature requirements for aquarium fish.
- You can now buy (dead) fish from the black market for dirt cheap.
- Last but now least, toilets are now valid fishing spots.

## Why It's Good For The Game
Fishing (and therefore aquariums) can feel a bit shallow at times, it has many cases of coder sprites in it, a few issues, things that weren't exactly implemented yet (like required temperature). It has its untapped potential though, one that doesn't necessarily involve getting gamer gear. Now, sharing some screenshots of the resprited fishing equipment:

and an example of soulless-ness: a holodeck crab.

## Changelog

:cl:
add: Expanded the fish breeding and traits system: fish can now pass down traits, size and weight to offspring. Some fishes have also a chance to generate new kind of fishes if conditions are met.
add: Added a fish analyzer that can be either bought from cargo, or printed if the advanced fishing tech node has been discovered.
add: The number of fillets a fish can make, its item size or the reagents you can get from blending it (yes, you can blend it now) now scale with fish size and weight. Fish cases also change size with the fish inside it because of it.
add: Added new fishing spots, with new fish and their own traits. This includes toilets.
Add: Added a new telescopic fishing rod as a goodie, and two new hooks, one from the adv. fishing tech node and another from the black market, as well as a case containing (dead) fish.
balance: Fish feed cans no longer can feed fishes indefinitely, but contain more feed now to compensate.
image: resprited fishing rods, hooks and the worm bait.
balance: It should be easier to fish people out of chasms even without a ruscue hook now, as long as you're in proximity of the chasm they fell in.
fix: You can now examine things, or interact with objects while fishing if the fishing rod isn't in your active hand.
qol: more accurate balloon messages for when fishing is interrupted.
/:cl:

### CODE REVIEW

**Comment by carlarctg** [comment]
I tried to fish up a corpse from lavaland once without a rescue hook, using a completed (?) bone fishing rod and dug-up worms for bait. After ten minutes I gave up, this is probably out of scope but in case it's not maybe you could take a look at that, because all these additions are great and all, but this is probably how most people will end up interacting with fishing.

**Comment by Ghommie** [comment]
Sure, I'll give it a look later.

**Comment by Y0SH1M4S73R** [comment]
We are getting ever closer to fish doctor becoming a real job.

**Comment by atlasle** [comment]
is that crab resprite included? it looks good

**Comment by Sealed101** [comment]
how could you not include new fish sprites bro smh!!!

cutie

File: code/modules/fishing/sources/source_types.dm [hunk 1]
```diff
@@ -65,11 +66,86 @@
...
+	fish_table = list(
+		FISHING_DUD = 5,
+		/obj/item/fish/chasm_crab/ice = 15,
+		/mob/living/simple_animal/hostile/asteroid/lobstrosity = 1,
+		/obj/effect/decal/remains/plasma = 1,
+		/obj/item/stack/ore/plasma = 3,
+		/obj/item/coin/plasma = 3,
+
+	)
+	fish_table = list(
```
Comment (AnturK) [inline code/modules/fishing/sources/source_types.dm:80]: I think you meant fish_counts here

File: code/modules/reagents/chemistry/holder.dm [hunk 2]
```diff
@@ -423,20 +423,27 @@
  * Reagent takes a PATH to a reagent.
  * Amount checks for having a specific amount of that chemical.
  * Needs matabolizing takes into consideration if the chemical is matabolizing when it's checked.
+ * Check types controls whether it should it should also include subtypes: ispath(type, reagent) versus type == reagent.
  */
-/datum/reagents/proc/has_reagent(reagent, amount = -1, needs_metabolizing = FALSE)
+/datum/reagents/proc/has_reagent(reagent, amount = -1, needs_metabolizing = FALSE, check_types = FALSE)
```
Comment (AnturK) [inline code/modules/reagents/chemistry/holder.dm:428]: why not check_subtypes ?

**Comment by Ghommie** [comment]
> how could you not include new fish sprites in the PR body bro smh!!!
> 
> cutie

IT WAS LATE AND I WAS TIRED!

Also, to @carlarctg, I've found the problem that made fishing bodies so damn hard: the pool of objects you can get from fishing on a chasm includes everything that fell in all chasms around the game world, not just nearby ones. Now, this has been changed to only pool objects from nearby chasms.

**Review by AnturK — COMMENTED** [review]
Looks very nice on quick glance

**Comment by Jacquerel** [comment]
> The pool of objects you can get from fishing on a chasm includes everything that fell in all chasms around the game world, not just nearby ones. Now, this has been changed to only pool objects from nearby chasms.

This was intentional design such that you didn't need to hunt down the specific chasm someone fell into, however if you feel it is a broad improvement I am not averse to changing it. I believe that GPS signal _should_ still function inside a chasm, so any lost miner should be findable.
Frankly I wouldn't be against a more weighted approach to simply make dead players their own list and have them as one of the most high priority results available, even without a rescue hook (which could change to a consumable forcing a guaranteed result?)

**Review by Jacquerel — CHANGES_REQUESTED** [review]
You're doing god's work and this is mostly just me being picky.

Nesting niche features like breeding and evolution on top of content that doesn't always even get used by one person per round is unironically the soul of ss13 development.

**Comment by Ghommie** [comment]
> This was intentional design such that you didn't need to hunt down the specific chasm someone fell into, however if you feel it is a broad improvement I am not averse to changing it. I believe that GPS signal _should_ still function inside a chasm, so any lost miner should be findable. Frankly I wouldn't be against a more weighted approach to simply make dead players their own list and have them as one of the most high priority results available, even without a rescue hook (which could change to a consumable forcing a guaranteed result?)

Thanks for the feedback. I will go on and check that the miner gps works inside chasms (it should already but still...) and perhaps give dead players their own lists as you say. Maybe I should also make the rescue hook work from anywhere, to justify its steep price.

File: code/datums/elements/skill_reward.dm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+/datum/element/skill_reward/proc/on_attack_hand(datum/source, mob/living/user, list/modifiers)
+	SIGNAL_HANDLER
+	if(!LAZYACCESS(modifiers, CTRL_CLICK) && !check_equippable(user)) //Allows other players to drag it around at least.
+		to_chat(user, span_warning("You feel completely and utterly unworthy to even touch \the [source]."))
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/datum/element/skill_reward/proc/on_equipped(datum/source, mob/living/user, slot)
+	SIGNAL_HANDLER
+	addtimer(CALLBACK(src, PROC_REF(drop_if_unworthy), source, user), 0) //Fire it on the next tick to avoid concurrency issues.
```
[the remark this answers is not in the document]
Comment (Jacquerel) [reply code/datums/elements/skill_reward.dm:29]: If it's not possible without a lot of work then we'll have to live with it (I've done it enough at work 😅) but scheduling events like this using timers is a recipe for real headscratcher bugs in the future.

[the remark this answers is not in the document]
Comment (Ghommie) [reply code/datums/elements/skill_reward.dm]: Ok, I've had to tweak the inventory code a little, but it's been done as you suggested!

**Review by Wallemations — APPROVED** [review]
Absolutely chumming at the sight of those sprites

**Comment by Ghommie** [comment]
Ok, so all that is left now is the icon for the achievement (and a few local tests).

**Comment by Ghommie** [comment]
Ok, it's taken some time but now it's ready and tested for the largest part.

**Comment by Jacquerel** [comment]
image
This might take me more than one lunch break

**Review by Wallemations — APPROVED** [review]
Those sprites are immaculate

I have one unofficial request and that's to add an achievement for fishing up a player body from a chasm and call it Bottomless Pit Inspector

File: code/__HELPERS/atmospherics.dm [hunk 4]
```diff
@@ -171,3 +171,34 @@ GLOBAL_LIST_EMPTY(gas_handbook)
...
+ * * * if the assoc list is null, then it'll be considered a safe gas and won't return FALSE.
+ * * extraneous_gas_limit: If a gas not in gases is found, this is the limit above which the proc will return FALSE.
+ */
+/proc/check_gases(list/gases, list/gases_to_check, extraneous_gas_limit = 0.1)
+	gases_to_check = gases_to_check.Copy()
+	for(var/id in gases)
+		var/gas_moles = gases[id][MOLES]
+		if(!(id in gases_to_check))
+			if(gas_moles > extraneous_gas_limit)
+				to_chat(world, "f1 [id]")
```
Comment (AnturK) [inline code/__HELPERS/atmospherics.dm:190]: You left some to_chat(worlds in here and below

File: code/datums/storage/subtypes/fish_case.dm [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+		return FALSE
+	if(user && (resolve_parent in user.get_equipped_items(TRUE)) && !user.is_holding(resolve_parent))
+		if(messages)
+			to_chat(user, span_warning("Either hold [resolve_parent] or place it nearby first!"))
+		return FALSE
+
+/datum/storage/fish_case/handle_enter(obj/item/storage/fish_case/source, obj/item/arrived)
+	. = ..()
+	if(istype(arrived))
+		source.w_class = arrived.w_class
```
Comment (AnturK) [inline code/datums/storage/subtypes/fish_case.dm:25]: These could possibly check if the storage is somewhere it can't be after w_class change.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/datums/storage/subtypes/fish_case.dm (hunk 5 above)
Comment (Ghommie) [reply code/datums/storage/subtypes/fish_case.dm]: Done.

File: code/modules/mob/inventory.dm [hunk 6]
```diff
@@ -163,15 +163,16 @@
 	I.forceMove(src)
 	held_items[hand_index] = I
 	SET_PLANE_EXPLICIT(I, ABOVE_HUD_PLANE, src)
-	I.equipped(src, ITEM_SLOT_HANDS)
```
Comment (AnturK) [inline code/modules/mob/inventory.dm:166]: Why did you move this after the visual update ?

**Review by AnturK — CHANGES_REQUESTED** [review]
Just few minor things and it should be good to go
